Bonding curve ratios

Hussain-7 opened the pull request.

Hussain-7 committed:
stable bonding curve getInputPrice() added

Hussain-7 committed:
issue with some formula

Hussain-7 committed:
New implmentation added working now for every case

Hussain-7 committed:
feat(stableSwap): getInputPrice() working

Hussain-7 committed:
chore(extra code): removed unnecessary code

Hussain-7 committed:
refactor(filename): file name changed

Hussain-7 committed:
feat(getOutputPrice): new function Implmented

Hussain-7 committed:
Finalized test and input price function

Hussain-7 committed:
cleaned the code removed unnecessary variables.

Hussain-7 committed:
Shifting code to amounts

Hussain-7 committed:
feat(AmountMath): ratio conversion added

Hussain-7 committed:
docs(eqs) : explained a bit of sub eqs used for y

Hussain-7 committed:
chore(code) : unessasary statements removed

Hussain-7 committed:
chore(eqs): comments for sub eqs in D calculation

Hussain-7 committed:
feat(Updated): Pushing latest stable bonding curve

Hussain-7 committed:
revert changes

Hussain-7 committed:
chore(stableSwap) : Shift curve logic to new dir

Hussain-7 committed:
refactor(formula) : using non simplified eq for d

Hussain-7 committed:
refactor(formula) : refactored formuala for y

Hussain-7 committed:
refactor(files): removed unnecessary files

Hussain-7 committed:
refactor(variables) : switched const to let

Hussain-7 committed:
refactor(file) : removed repeatition of functions

Hussain-7 committed:
feat(test) : Unit tests for d added

Hussain-7 committed:
refactor(vars) : variables refactored to standard

Hussain-7 committed:
fixes(vars): minor fixes

Hussain-7 committed:
feat(D): ratios implemented in formula of D

Hussain-7 committed:
feat(ratio): conversions completed

Hussain-7 committed:
fixes(unitTest) : Unit tests updated

Hussain-7 committed:
refactor(UnitTest): UnitTest for d

Hussain-7 committed:
documentation(stableSwap) : comments added

Hussain-7 committed:
fix(ratio): Issue in ratio

Hussain-7 committed:
feat(test) : More test added

Chris-Hibbert commented on packages/zoe/src/contractSupport/safeMath.js:
```diff
@@ -19,5 +19,12 @@ export const natSafeMath = harden({
+  power: (x, y) => {
+    let result = 1n;
+    for (let i = 0; i < y; i++) {
+      result *= Nat(x);
```
why repeated multiplication rather than just using `x**y`?

Hussain-7 replied:
Actually this is not needed now will remove

Chris-Hibbert commented on packages/run-protocol/test/amm/stableSwap/test-stableSwap-dvalue.js:
```diff
@@ -0,0 +1,89 @@
+const logResults = (poolValues, output) => {
+  console.log('\npool :', poolValues);
+  console.log('D :', output);
+};
```
console logging during tests doesn't belong in the final check-in.

This doesn't have to be removed now, but definitely before it is done.

Hussain-7 replied:
got it

Chris-Hibbert commented on .gitignore:
```diff
@@ -63,3 +63,7 @@ bundle-*.js
 /packages.png
 packages/ui-components/compiled
 packages/ui-components/dist
+packages/zoe/src/contractSupport/stableBondingCurves1.js
+packages/zoe/src/contractSupport/stableBondingCurves2.js
+packages/zoe/src/contractSupport/stableBondingCurves.js
+packages/zoe/test/unitTests/contractSupport/test-stableBondingCurves.js
```
I think if you don't check in changes to .gitignore your local copy will still work to prevent these other files from being added to the commit. We certainly don't want these added to the file.

Hussain-7 replied:
Yes sure will remove them

Chris-Hibbert committed:
refactor: drop Ratios, cleanup, simplify

Ratios and Amount get in the way here, since we have to mutliply
values from different brands. You were right initially to omit them.

We only need to convert to Amounts in
getStableInputPrice/getStableOutputPrice.

getD() works, and `test-stableSwap-dvalue` passes.

getY() is confused, sorry, I ran out of time to finish fixing it.
As I understand it, d.denominator shouldn't be necessary, since getD()
returns a bigInt. the yNum and yDen in this branch are wrong, since
they're producing a denominator much larger than the numerator, which
makes the floorDivide on line 160 return zero, so
test-stableSwap-price fails.  But it should follow the model in getD,
of dealing with bigInts.

Having switched to bigInts, calculateSwap can be much simpler. If I
ask you to do something that results in making Amounts with a dummy
brand, then I've made a mistake.

I did some cleanups and simplifications for style in
getStableInputPrice. Once getY() is fixed, that should work.

getStableOutputPrice doesn't seem right to me. Once we have
getStableInputPrice working and the tests reinstated, we can take
another look, but I'm very sceptical about getting that answer by
successive approimation.

The tests for getD() got much simpler since it's not dealing with
Amounts any more.

I switched to triple equals. JavaScript has some broken edges, and `==`
and `!=` are deprecated.

I also replaced `++` with `+= 1`, since that's what our style guide
requires.

I renamed a few things using camelCase

Hussain-7 committed:
Merge pull request #5 from Chris-Hibbert/bonding-curve-ratios

refactor: drop Ratios, cleanup, simplify
Have ran the code and tested everything works as chris told so merging

Hussain-7 committed:
fix(getY): fixed issues with getY function

Hussain-7 committed:
minor changes

Chris-Hibbert commented on packages/run-protocol/test/amm/stableSwap/test.js:
```diff
@@ -0,0 +1,5 @@
+// @ts-check
+
+const numerator = 12150641942309009447276218899927731304676524131073406481203200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000n;
```
This doesn't look like a test to check in.

Hussain-7 replied:
Will remove it.

Chris-Hibbert commented on packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js:
```diff
@@ -0,0 +1,299 @@
+  );
+  const inputAmountWithoutFeeCut = inputAmount;
+  const poolValues = poolAmounts.map(amount => amount.value);
+  const poolBrands = poolAmounts.map(amount => amount.brand);
```
Nothing uses poolBrands. Don't create it or return it.

Hussain-7 replied:
yes will remove it

Chris-Hibbert commented on packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js:
```diff
@@ -0,0 +1,299 @@
+    inputAmount.brand,
+  );
+  // Fee ratio multiplied by inputAmount to get inputAmount After fee cut
+  let inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
+  // Normalizing input amount according to pool value
+  inputAmountAfterFeeCut = AmountMath.make(
+    inputAmountAfterFeeCut.brand,
+    inputAmountAfterFeeCut.value,
+  );
```
```suggestion
  const inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
```

Hussain-7 replied:
Using the same variable below in `line 61` so can't make it const.

Chris-Hibbert replied:
Line 61 makes an Amount from the components of an Amount. It's completely unnecessary. The value computed on line 59 (and reused on line 51) is all that's needed, so it can be `const`.

Chris-Hibbert commented on packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js:
```diff
@@ -0,0 +1,299 @@
+    inputAmountAfterFeeCut.brand,
+    inputAmountAfterFeeCut.value,
+  );
+  const inputAmountWithoutFeeCut = inputAmount;
```
This variable isn't useful.

Hussain-7 replied:
removed

Chris-Hibbert commented on packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js:
```diff
@@ -0,0 +1,299 @@
+
+  return {
+    inputAmountAfterFeeCut,
+    inputAmountWithoutFeeCut,
```
```suggestion
    inputAmountWithoutFeeCut: inputAmount,
```

Chris-Hibbert commented on packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js:
```diff
@@ -0,0 +1,299 @@
+    poolValues,
+  );
+  const outputBrand = poolAmounts[tokenIndexTo].brand;
+  const priceRatio = makeRatioFromAmounts(
```
We don't need priceRatio.

Hussain-7 replied:
alright

Chris-Hibbert commented:
Now that `getD` and `getY` are working well, I can see other things that need some attention.

`commonStablePrice` isn't actually common between input and output, but parts of it can be. I think what we want to keep is the assertions and building poolValues. The fees are different for input and output.

The way we actually want to charge fees is to subtract it from the input before `getY` for `getStableInputPrice()` and add it to the output of `getY` for `getStableOutputPrice()`. I think`getY` doesn't care whether Y is growing or shrinking.

```
export const getStableInputPrice = async (
  inputAmount,
  tokenIndexFrom,
  tokenIndexTo,
  poolAmounts,
  feeBasisPoints = 30n,
) => {
  const inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
  const x = inputAmountAfterFeeCut.value + poolValues[tokenIndexFrom];
  const y = getY(x, tokenIndexFrom, tokenIndexTo, poolValues);
  const swapResult = poolValues[tokenIndexTo] - y;
  return {
    inputAmount,
    outputAmount: AmountMath.make(outputBrand, swapResult),
  };
```

```
export const getStableOutputPrice = async (
  outputAmount,
  tokenIndexFrom,
  tokenIndexTo,
  poolAmounts,
) => {
  const x = outputAmount.value + poolValues[tokenIndexFrom];
  const y = getY(x, tokenIndexFrom, tokenIndexTo, poolValues);
  const swapResult = poolValues[tokenIndexTo] - y;
  const swapResultPlusFee = floorMultiplyBy(swapResult, addFeeRatio);

  return {
    inputAmount: AmountMath.make(inputBrand, swapResultPlusFee),
    outputAmount,
  };
```

I haven't verified that I got everything in the right place or tested, this is just to show the approach I think will work.

Hussain-7 committed:
chore(gitignore): removed some files in gitignore

Hussain-7 committed:
chores(file): unessasary file removed

Hussain-7 committed:
fixed Issue mentioned in pr

Hussain-7 committed:
price ratio calculation removed

Hussain-7 commented:
> Now that `getD` and `getY` are working well, I can see other things that need some attention.
> 
> `commonStablePrice` isn't actually common between input and output, but parts of it can be. I think what we want to keep is the assertions and building poolValues. The fees are different for input and output.
> 
> The way we actually want to charge fees is to subtract it from the input before `getY` for `getStableInputPrice()` and add it to the output of `getY` for `getStableOutputPrice()`. I think`getY` doesn't care whether Y is growing or shrinking.
> 
> ```
> export const getStableInputPrice = async (
>   inputAmount,
>   tokenIndexFrom,
>   tokenIndexTo,
>   poolAmounts,
>   feeBasisPoints = 30n,
> ) => {
>   const inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
>   const x = inputAmountAfterFeeCut.value + poolValues[tokenIndexFrom];
>   const y = getY(x, tokenIndexFrom, tokenIndexTo, poolValues);
>   const swapResult = poolValues[tokenIndexTo] - y;
>   return {
>     inputAmount,
>     outputAmount: AmountMath.make(outputBrand, swapResult),
>   };
> ```
> 
> ```
> export const getStableOutputPrice = async (
>   outputAmount,
>   tokenIndexFrom,
>   tokenIndexTo,
>   poolAmounts,
> ) => {
>   const x = outputAmount.value + poolValues[tokenIndexFrom];
>   const y = getY(x, tokenIndexFrom, tokenIndexTo, poolValues);
>   const swapResult = poolValues[tokenIndexTo] - y;
>   const swapResultPlusFee = floorMultiplyBy(swapResult, addFeeRatio);
> 
>   return {
>     inputAmount: AmountMath.make(inputBrand, swapResultPlusFee),
>     outputAmount,
>   };
> ```
> 
> I haven't verified that I got everything in the right place or tested, this is just to show the approach I think will work.

### Issues in approach mentioned for getStableOutputPrice:
- The approach you mentioned, I have already tried that before but this has a problem. That is if we use the result of getOutputPrice function to verify it with getInputPrice(). The amount of token we get if use the output Amount with getInputPrice function are lesser that what user ask for.
- The test below will fail and we get Input Price as `144n`.
- If we use `144n` with get stablePrice() to verify result .The output value for token A we get `972n` <`1000n` hence wrong.
<img src="https://user-images.githubusercontent.com/55734494/160941994-a2084de8-9466-4303-8ea0-fddaa138024f.png"  width="350" height="300" center/>

- The images are just reference to explain that this does not indicate any issue with getY() function its the logic of getY() function that is like this. Result from web as well indicate same thing.
- Like if we want to swap `1000` token `A` to `B`. Then we get `101.93763` of Token `B`.
- But If we do vice versa and swap `101.93763` Token `B` to Token `A`. Then we get `983.28669` of Token `A`.
- This is the reason I was increasing token B amount iteratively unless it give amount of token A that is greater than output amount needed.
<img src="https://user-images.githubusercontent.com/55734494/160939020-500a2b92-3a3c-49b5-99ba-2359a2791caf.png"  width="350" height="300"/>
<img src="https://user-images.githubusercontent.com/55734494/160939030-aec86089-1ef1-428d-bd55-27b8dd6a79c5.png"  width="350" height="300"/>

Hussain-7 committed:
getStableInputPrice refactored a bit

Hussain-7 committed:
commonStablePrice function refactored

Chris-Hibbert commented:
I'll take a closer look, but I suspect the difference is that we're charging fees. The tool you show there doesn't include fees, does it? This means that getInputPrice and getOutputPrice aren't expected to give the same result. 

And that the tests shouldn't get the same result as the tools do unless the fees are set to zero.

Hussain-7 commented:
> I'll take a closer look, but I suspect the difference is that we're charging fees. The tool you show there doesn't include fees, does it? This means that getInputPrice and getOutputPrice aren't expected to give the same result. 
> 
> And that the tests shouldn't get the same result as the tools do unless the fees are set to zero.

- They are also charging a fee which is equal to 40 basis point in our case
- Where as we have fee set to 30 basis point

Chris-Hibbert commented:
If we wanted getStableInputPrice and getStableOutputPrice that computed the same thing, we wouldn't need two functions. They're supposed to be doing different things. getInputPrice(X) asks "What would I get for X?", while getOutputPrice(Y) asks "What would I have to pay to get Y?"

We charge a fee on getInputPrice before getting the results from calculateSwap. With getOutputPrice we charge the fee on the amount output from the pool. This ensures that the pool calculation is based on the number the user named.

If the price is currently .95 USDT per BLD, and we're charging a 50 BP fee, then

getInputPrice(1000 BLD)    =>  945 USDT    (== (1000 - 5) * .95)
getInputPrice(1000 USDT) => 1047 BLD    (== (1000 - 5) / .95)

getOutputPrice(1000 BLD)   => 955 USDT   (== (1000 * .95) + 5)
getOutputPrice(1000 USDT) => 1057 BLD  (== (1000 / .95) + 5)

if swapResult in getStableOutputPrice is used as the inputPrice in a call to getStableInputPrice(), It's effectively asking "what would I get for 955 USDT?" rather than charging a fee on the 950 that we've established is what needs to be paid in order to get 1000 BLD.

Hussain-7 commented:
- @Chris-Hibbert I totally understand your point. Even I was thinking what's the point of `getOutputPrice()` in this case.
- The problem is the getY() and getD() calculation do not work like for price calculation in constant product.
- Hence if we use the above explained process we don't get the right value for getOutputPrice().
- As we know the `stableCurve` take into account both the `input` or the `output amount` as well as the `pool` amounts of each token. So there is no such thing as a fixed price between two tokens. 
- Hence my point is either we would need to change the `getY()` and `getD()` calculations. Which were derived from the stableSwap White paper. Or there has to be some way to use them to find outputPrice which clearly I cannot see.
- I have tried every possible way to find `outputPrice` through the method you stated but thats returning wrong value. 
- So currently I have no clue how to proceed the `getOutputPrice` calculation.
